src/main/kotlin/org/wfanet/panelmatch/client/privatemembership/JniQueryEvaluator.kt

```diff
@@ -14,29 +14,129 @@
+import com.google.privatemembership.batch.server.RawDatabaseShardKt.bucket
+import com.google.privatemembership.batch.server.Server.RawDatabaseShard
+import com.google.privatemembership.batch.server.SumCiphertextsRequestKt.summation
+import com.google.privatemembership.batch.server.applyQueriesRequest
+import com.google.privatemembership.batch.server.finalizeResultsRequest
+import com.google.privatemembership.batch.server.rawDatabaseShard
+import com.google.privatemembership.batch.server.sumCiphertextsRequest
+import com.google.protobuf.ByteString
+import rlwe.Serialization.SerializedSymmetricRlweCiphertext
 
 /** [QueryEvaluator] that calls into C++ via JNI. */
 class JniQueryEvaluator(private val parameters: QueryEvaluatorParameters) : QueryEvaluator {
+  private val privateMembershipParameters: Parameters by lazy {
+    Parameters.parseFrom(parameters.serializedPrivateMembershipParameters)
+  }
+
+  private val privateMembershipPublicKey: PublicKey by lazy {
+    PublicKey.parseFrom(parameters.serializedPublicKey)
+  }
 
   override fun executeQueries(
     shards: List<DatabaseShard>,
     queryBundles: List<QueryBundle>
   ): List<Result> {
-    TODO()
+    val request = applyQueriesRequest {
+      parameters = privateMembershipParameters
+      publicKey = privateMembershipPublicKey
+      finalizeResults = false
+      queries += queryBundles.map(QueryBundle::encryptedQueries)
+      rawDatabase =
+        rawDatabase {
+          this.shards += shards.map(DatabaseShard::toPrivateMembershipRawDatabaseShard)
+        }
+    }
+
+    val response = JniPrivateMembership.applyQueries(request)
+
+    return response.queryResultsList.map { encryptedQueryResult ->
+      resultOf(
+        queryIdOf(encryptedQueryResult.queryMetadata.queryId),
+        encryptedQueryResult.toByteString()
+      )
+    }
   }
 
   override fun combineResults(results: Sequence<Result>): Result {
-    TODO()
+    val sums = mutableListOf<MutableList<SerializedSymmetricRlweCiphertext>>()
+    var firstQueryMetadata: QueryMetadata? = null
+    val queryIds = mutableSetOf<Int>()
+
+    for (result in results) {
+      val encryptedQueryResult =
+        EncryptedQueryResult.parseFrom(result.serializedEncryptedQueryResult)
+      for ((i, ciphertext) in encryptedQueryResult.ciphertextsList.withIndex()) {
+        if (i == sums.size) sums += mutableListOf<SerializedSymmetricRlweCiphertext>()
+        sums[i].add(ciphertext)
+      }
+
+      if (firstQueryMetadata == null) firstQueryMetadata = encryptedQueryResult.queryMetadata
+
+      check(result.queryId.id == encryptedQueryResult.queryMetadata.queryId)
+      queryIds.add(result.queryId.id)
+    }
+
+    require(queryIds.size == 1) { "Mismatching query ids" }
+    requireNotNull(firstQueryMetadata) { "Cannot combine zero results" }
+
+    if (sums.isEmpty()) {
+      return resultOf(queryIdOf(firstQueryMetadata.queryId), ByteString.EMPTY)
```

Comment:
This if statement is enacted if either results is empty or each entry of results is empty. I think it will be clearer if you did this check at the top of the function.